Restrict editor programs

zlavergne opened the pull request:
**Requires a db migration**
This feature allows for project managers to restrict the editor programs that can be used during either editing stage.

* Added `map_editors` and `validate_editors` columns and accompanying dto's

* Added list of editors in the `Settings` menu of a project-edit page where a project manager can select permitted editors for each editing stage
![Screen Shot 2019-03-21 at 10 37 50 AM](https://user-images.githubusercontent.com/9849118/54944416-10f7a880-4ef9-11e9-8824-6867463ebf3c.png)


* Updated the editor dropdown for to populate only with the permitted editors
![Screen Shot 2019-03-25 at 12 28 35 PM](https://user-images.githubusercontent.com/9849118/54944636-90857780-4ef9-11e9-9797-c8bd7a727e72.png)


> Note: iD X is added as an editor option. iD X will be the iD editor to support FB's ML as a Service project. It will be implemented in a future commit.

* only `ADMIN`s can add `iD X` as an editor program

* only user with `expertMode` enabled can use the `iD X` editor

michael63-osm committed:
Restrict PM promotion to admin

zlavergne commented:
Reset the branch in order to remove `id_x` implementation

gaoxm commented:
@zlavergne the changes look good overall! I have a couple of ideas about the eng design (that may help simplify the changes) for discussion:
(1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.
(2) How much work will it be for TM admin to do the DB migration, given that they'll need to populate the value for all existing projects? If my assumption for (1) is true, then is it possible to just add a "id_x_enabled=true" type of metadata to some existing project column that contains a JSON blob? No DB migration will be needed that way, and we can default it to false when the property is not there (e.g. for the existing projects) :)

@jeffdefacto @chrisklaiber what do you guys think?

zlavergne commented:
> (1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.

I did it this way after [this discussion on Slack](https://hotosm.slack.com/archives/CH19F91EU/p1553103960011100) with @jeffdefacto and Felix. Let me know if we want to go a different route.

> (2) How much work will it be for TM admin to do the DB migration, given that they'll need to populate the value for all existing projects? If my assumption for (1) is true, then is it possible to just add a "id_x_enabled=true" type of metadata to some existing project column that contains a JSON blob? No DB migration will be needed that way, and we can default it to false when the property is not there (e.g. for the existing projects) :)

Good catch. I was working with a fresh db and hadn't thought about that. I updated the migrations in [this commit](https://github.com/KaartGroup/tasking-manager/pull/14/commits/f1af4a0d3eaa4d87d9870a2716cdc69180009e86) to take this into account [with these lines.](https://github.com/KaartGroup/tasking-manager/blob/f1af4a0d3eaa4d87d9870a2716cdc69180009e86/migrations/versions/d01c4ceddeb4_.py#L24-L25) Now when alembic does the migration, it will add all the regular editors so they don't lose any functionality.

zlavergne commented:
Circle is failing due to the debian version. See [#1141 on hotosm](https://github.com/hotosm/tasking-manager/pull/1411)

jeffdefacto commented:
> (1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.

I think it makes sense to have them as a separate option. The feature is expanding upon only allowing JOSM for validation so I think to remain consistent we should allow the turning on or off of any editor for validation. For our workflow we don't use iD for the validation stage when using the service so turning it off for simplicity could be nice.

gaoxm commented:
> > (1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.
> 
> I did it this way after [this discussion on Slack](https://hotosm.slack.com/archives/CH19F91EU/p1553103960011100) with @jeffdefacto and Felix. Let me know if we want to go a different route.
> 
> > (2) How much work will it be for TM admin to do the DB migration, given that they'll need to populate the value for all existing projects? If my assumption for (1) is true, then is it possible to just add a "id_x_enabled=true" type of metadata to some existing project column that contains a JSON blob? No DB migration will be needed that way, and we can default it to false when the property is not there (e.g. for the existing projects) :)
> 
> Good catch. I was working with a fresh db and hadn't thought about that. I updated the migrations in [this commit](https://github.com/KaartGroup/tasking-manager/pull/14/commits/f1af4a0d3eaa4d87d9870a2716cdc69180009e86) to take this into account [with these lines.](https://github.com/KaartGroup/tasking-manager/blob/f1af4a0d3eaa4d87d9870a2716cdc69180009e86/migrations/versions/d01c4ceddeb4_.py#L24-L25) Now when alembic does the migration, it will add all the regular editors so they don't lose any functionality.

Sorry I missed the earlier discussion on Slack. Will review the updates ASAP.

gaoxm commented on client/app/admin/edit-project/edit-project.controller.js:
```diff
@@ -40,6 +40,22 @@
         };
 
+        // Editors for mapping
+        vm.mapEditors = {
```
minor naming suggestion: maybe just use "mappingEditors"? "mapEditors" sounds more like a human role rather than a tool.

gaoxm commented on client/app/admin/edit-project/edit-project.controller.js:
```diff
@@ -40,6 +40,22 @@
+        };
+
+        // Editors for validation
+        vm.validateEditors = {
```
similarly, use "validationEditors"?

gaoxm commented on client/app/admin/edit-project/edit-project.controller.js:
```diff
@@ -167,6 +187,8 @@
             if (!requiredFieldsMissing && vm.editForm.$valid){
                 vm.project.mappingTypes = getMappingTypesArray();
+                vm.project.mapEditors = getMapEditorsArray();
+                vm.project.validateEditors = getValidateEditorsArray();
```
same

gaoxm commented on client/app/admin/edit-project/edit-project.controller.js:
```diff
@@ -928,6 +952,89 @@
+            if (vm.project.mapEditors) {
+                if (vm.project.mapEditors.indexOf("ID") != -1) {
+                    vm.mapEditors.id = true;
+                }
```
maybe a little simpler this way?
  `vm.mapEditors.id = vm.project.mapEditors.indexOf("ID") != -1;`

gaoxm commented on client/app/admin/edit-project/edit-project.controller.js:
```diff
@@ -928,6 +952,89 @@
+        /**
+         * Get map editors in array
+         */
+        function getMapEditorsArray(){
```
similarly, getMappingEditorsArray()

gaoxm commented on client/app/admin/edit-project/edit-project.controller.js:
```diff
@@ -928,6 +952,89 @@
+        /**
+         * Get validate editors in array
+         */
+        function getValidateEditorsArray(){
```
getValidationEditorsArray()

gaoxm commented on client/app/admin/edit-project/edit-project.controller.js:
```diff
@@ -928,6 +952,89 @@
+        function populateEditorsForValidation(){
+            if (vm.project.validateEditors) {
+                if (vm.project.validateEditors.indexOf("ID") != -1) {
+                    vm.validateEditors.id = true;
```
same

gaoxm commented on client/app/project/project.controller.js:
```diff
@@ -59,7 +59,10 @@
         vm.editorStartError = '';
-        vm.selectedEditor = 'ideditor';
+        vm.selectedMapEditor = 'ideditor';
+        vm.selectedValidateEditor = 'josm';
```
same: `selectedMappingEditor` and `selectedValidationEditor`

gaoxm commented on client/app/project/project.controller.js:
```diff
@@ -115,6 +119,7 @@
                 var resultsPromise = accountService.getUser(session.username);
                 resultsPromise.then(function (user) {
                     vm.user = user;
+                    console.log(user);
```
remove

jeffdefacto commented on client/app/services/editor.service.js:
```diff
@@ -79,6 +80,40 @@
         }
 
+        /**
+         * Launch the iD X editor
```
Shouldn't the id X stuff be a separate PR?

zlavergne replied:
Missed that one. Fixed with 29d203a

gaoxm commented on migrations/versions/215392b11b0e_.py:
```diff
@@ -0,0 +1,44 @@
+    op.create_index('idx_project_info composite', 'project_info', ['locale', 'project_id'], unique=False)
+    op.create_index('textsearch_idx', 'project_info', ['text_searchable'], unique=False)
+    op.add_column('projects', sa.Column('map_editors', postgresql.ARRAY(sa.Integer()), nullable=True))
+    op.add_column('projects', sa.Column('validate_editors', postgresql.ARRAY(sa.Integer()), nullable=True))
```
same: `mapping_editors` and `validation_editors`

gaoxm commented on migrations/versions/215392b11b0e_.py:
```diff
@@ -0,0 +1,44 @@
+               existing_type=sa.INTEGER(),
+               nullable=False)
+    op.create_index(op.f('ix_projects_map_editors'), 'projects', ['map_editors'], unique=False)
+    op.create_index(op.f('ix_projects_validate_editors'), 'projects', ['validate_editors'], unique=False)
```
do we need indices on these columns? For use cases where you want to search for all projects with "iD X" enabled, does this index support that?

zlavergne replied:
Yes. I put that in so projects would be searchable based on editors in the future. I tried to mimic similar implementations.

gaoxm replied:
👍 sounds good!

gaoxm commented on migrations/versions/d01c4ceddeb4_.py:
```diff
@@ -0,0 +1,39 @@
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_index('idx_project_info composite', table_name='project_info')
+    op.create_index('idx_project_info composite', 'project_info', ['locale', 'project_id'], unique=False)
+    op.drop_index('idx_project_info composite', table_name='project_info')
```
why is it dropped again right after creation?

zlavergne replied:
I'm not entirely sure. That part was autogenerated by alembic. I'll remove the `drop`s. I can't think of any reason they would need to be there.

gaoxm commented on migrations/versions/d01c4ceddeb4_.py:
```diff
@@ -0,0 +1,39 @@
+    op.add_column('projects', sa.Column('map_editors', postgresql.ARRAY(sa.Integer()), nullable=False, server_default='{0,1,2,3}'))
+    op.add_column('projects', sa.Column('validate_editors', postgresql.ARRAY(sa.Integer()), nullable=False, server_default='{0,1,2,3}'))
+    op.create_index(op.f('ix_projects_map_editors'), 'projects', ['map_editors'], unique=False)
+    op.create_index(op.f('ix_projects_validate_editors'), 'projects', ['validate_editors'], unique=False)
```
same questions

gaoxm commented on server/models/dtos/project_dto.py:
```diff
@@ -105,6 +118,8 @@ class ProjectDTO(Model):
     task_creation_mode = StringType(required=True, serialized_name='taskCreationMode',
                                     validators=[is_known_task_creation_mode], serialize_when_none=False)
+    map_editors = ListType(StringType, serialized_name='mapEditors', validators=[is_known_editor])
+    validate_editors = ListType(StringType, serialized_name='validateEditors', validators=[is_known_editor])
```
same: `mapping_editors` and `validation_editors`

gaoxm commented on server/models/dtos/project_dto.py:
```diff
@@ -118,6 +133,8 @@ class ProjectSearchDTO(Model):
     text_search = StringType()
     is_project_manager = BooleanType(required=True, default=False)
+    map_editors = ListType(StringType, validators=[is_known_editor])
+    validate_editors = ListType(StringType, validators=[is_known_editor])
```
same

gaoxm commented on server/models/dtos/project_dto.py:
```diff
@@ -129,9 +146,17 @@ def __hash__(self):
+        if self.map_editors:
+            for map_editor in self.map_editors:
+                hashable_map_editors = hashable_map_editors + map_editor
+        hashable_validate_editors = ''
```
same: `hashable_mapping_editors` and `hashable_validation_editors`

zlavergne replied:
Fixed 23c4bb4

gaoxm commented on server/models/postgis/project.py:
```diff
@@ -64,6 +64,10 @@ class Project(db.Model):
 
+    # Editors
+    map_editors = db.Column(ARRAY(db.Integer), default=[0, 1, 2, 3], index=True, nullable=False)
+    validate_editors = db.Column(ARRAY(db.Integer), default=[0, 1, 2, 3], index=True, nullable=False)
```
same for naming; also, how about using the Editors Enum values below for setting the defaults?

zlavergne replied:
Do you mean just replacing the ints with `Editors[editor]` or making it an `ARRAY(Enum(Editors))`?

gaoxm replied:
the former, e.g. default=[Editors.ID.value, ...]

zlavergne replied:
Fixed 23c4bb4

gaoxm commented on server/models/postgis/project.py:
```diff
@@ -203,6 +207,17 @@ def update(self, project_dto: ProjectDTO):
+            map_editors_array.append(Editors[map_editor].value)
+        self.map_editors = map_editors_array
+
+        validate_editors_array = []
```
`mapping_editors_array` and `validation_editors_array`

zlavergne replied:
Fixed 23c4bb4

gaoxm commented on server/models/postgis/project.py:
```diff
@@ -369,6 +384,20 @@ def _get_project_and_base_dto(self):
+            for validate_editor in self.validate_editors:
+                validate_editors.append(Editors(validate_editor).name)
+
+            base_dto.validate_editors = validate_editors
```
same

zlavergne replied:
Fixed 23c4bb4

gaoxm reviewed (changes requested):
looks good overall! Some minor comments inline, mostly around naming

jeffdefacto commented on client/app/services/editor.service.js:
```diff
@@ -18,6 +18,7 @@
             launchFieldPapersEditor: launchFieldPapersEditor,
             launchPotlatch2Editor: launchPotlatch2Editor,
             launchIdEditor: launchIdEditor,
+            launchIdX: launchIdX,
```
Here too :)

zlavergne replied:
Argggghhhh. Search and Replace has failed me miserably

zlavergne replied:
Fixed 23c4bb4

zlavergne commented:
23c4bb4 fixes all review changes

gaoxm commented:
LGTM :)

pantierra committed:
Temporarily reestablish main Dockerfile for Elastic Beanstalk deployment

Commit:
Merge pull request #1437 from hotosm/fix-ci-again

Fix Elastic Beanstalk deployment

ethan-nelson committed:
Restrict PM promotion to admin

Commit:
Merge pull request #1443 from ethan-nelson/limit-pm-promotion

Restrict PM promotion to admin

Commit:
Restrict editor programs for mapping and validation stages

>**This PR includes a db migration**

Related Issue: #893

* Added list of editors in the `Settings` menu of a project-edit page where a project manager can select permitted editors for each editing stage
![Screen Shot 2019-03-21 at 10 37 50 AM](https://user-images.githubusercontent.com/9849118/54944416-10f7a880-4ef9-11e9-8824-6867463ebf3c.png)

* Updated the editor dropdown for to populate only with the permitted editors
![Screen Shot 2019-03-25 at 12 28 35 PM](https://user-images.githubusercontent.com/9849118/54944636-90857780-4ef9-11e9-9797-c8bd7a727e72.png)

zlavergne committed:
Merge branch 'develop' into restrict-editor-programs

Commit:
Restrict editor programs for mapping and validation stages

>**This PR includes a db migration**

Related Issue: #893

* Added list of editors in the `Settings` menu of a project-edit page where a project manager can select permitted editors for each editing stage
![Screen Shot 2019-03-21 at 10 37 50 AM](https://user-images.githubusercontent.com/9849118/54944416-10f7a880-4ef9-11e9-8824-6867463ebf3c.png)

* Updated the editor dropdown for to populate only with the permitted editors
![Screen Shot 2019-03-25 at 12 28 35 PM](https://user-images.githubusercontent.com/9849118/54944636-90857780-4ef9-11e9-9797-c8bd7a727e72.png)

zlavergne commented:
Closing this since it was merged into mainline way back when and has been piped into ours with the sync to mainline.
